[FIXED JENKINS-28926] Block while upstream/downstream building cycles never complete

stephenc opened the pull request:
See [JENKINS-28926](https://issues.jenkins-ci.org/browse/JENKINS-28926)
- One could argue that without this change the system is functioning correctly and that previous behaviour
  was a bug. On the other hand, people have come to rely on the previous behaviour.
- The issue really centeres around state changes in the blocked tasks. Since blocking on upstream/downstream
  relies on checking the building projects and the queued (excluding blocked) tasks we need any change in
  the blocked task list to be visible immediately (i.e. update the snapshot)
- I was able to reliably reproduce this behaviour with a convoluted set of manually configured projects
  but turning this into a test case has not proved quite as easy. Manual testing confirms that the issue is
  fixed for my manual test case
- I have also added a sorting of the blocked list when probing for tasks to unblock. This should prioritise
  tasks as intended by the QueueSorter

@reviewbybees

stephenc committed:
[FIXED JENKINS-28926] Block while upstream/downstream building cycles never complete

- One could argue that without this change the system is functioning correctly and that previous behaviour
  was a bug. On the other hand, people have come to rely on the previous behaviour.
- The issue really centeres around state changes in the blocked tasks. Since blocking on upstream/downstream
  relies on checking the building projects and the queued (excluding blocked) tasks we need any change in
  the blocked task list to be visible immediately (i.e. update the snapshot)
- I was able to reliably reproduce this behaviour with a convoluted set of manually configured projects
  but turning this into a test case has not proved quite as easy. Manual testing confirms that the issue is
  fixed for my manual test case
- I have also added a sorting of the blocked list when probing for tasks to unblock. This should prioritise
  tasks as intended by the QueueSorter

jtnord commented:
I'm not sure how this fixes the linked issue.

From what I can tell it will prevent issues where multiple jobs should be blocked by each other are all in the queue, but if none are running all can start simultaneously.  But I don't see how it fixes the issue that when a job stops that any blocked jobs will then be released.

Can you elaborate some more on this?

Otherwise I am :+1: on the code changes.

stephenc committed:
[JENKINS-28926] Adding test case

- I was forgetting the call to `rebuildDependencyGraph()` which was why the test didn't work for me

oleg-nenashev commented:
:+1: 
I suspect there will be an additional performance degradation.  Probably it makes sense to refactor <code>Queue::maintain()</code> in order to avoid nasty cycles of wasted efforts:

Example for parallel builds depending on the same node label
0) e.g. a required node gets connected. Suppose it has several executors (e.g. 4), but 1000 builds are waiting for them
1) We unblock multiple jobs at once. For each one we call updateSnapshot
2) We sort buildables and update the snapshot again
3) We start assigning items to labels. 
- If the build can be built, we assign it and update the snapshot (4 times)
- If not, we move it back to blocked and ... update the snapshot again (996 times)
4) ....
5) 1 build finishes
6) We make 996 items buildable and iterate again :(

stephenc commented:
@jtnord so there are two points where we check if a job is blocked (and where the job being blocked actually matters... we can check in other places but those places do not have any affect on building)

The first is the iteration of the blocked list. Upstream/downstream blockage is determined based on the items that are building plus all the **non-blocked** items in the queue. So what you had is all the items are "blocked" by each other, we check them all in turn, and as none of the other "blocking" projects are building or in the queue (excluding blocked items!) we promote **all** of them to buildable...

The second is the iteration of the buildable items. So here we would update the snapshot first... exposing the live state of what is in the queue, and then check each item in turn... of course now they see that all their buddies are in the queue (excluding blocked items) so all of them are blocked and we would put them back into the blocked list.

The changes here are two fold.

First, when processing the blocked to unblocked transition, every time we move one of the blocked ones out of blocked and into buildable, we now update the snapshot. Thus one and only one of the blocked chain will be moved from blocked to buildable in that iteration.

Second, when processing the buildable to started transition, every time we move one of the buildable ones back to blocked we update the snapshot so that if there are subsequent "buddies" from the cycle in the buildable state, they will indeed be able to start.

stephenc commented:
@oleg-nenashev I believe you misunderstand how this change affects the queue processing. There should be minimal impact as actually what is happening currently is the entire cycle of blocked projects is jumping between blocked and buildable continuously.

With this change we promote one of the cycle up to buildable at a time. 

The assignment of labels does not affect the blocked state. A job is buildable if waiting for a label. It is blocked by other jobs or external conditions not the absence of a node with a label to pick up the job

oleg-nenashev commented:
I'm OK wit the change, my concern is related to the overall behavior

> The assignment of labels does not affect the blocked state. A job is buildable if waiting for a label. It is blocked by other jobs or external conditions not the absence of a node with a label to pick up the job

We can replace the example by any Queue dispatcher like Throttle Concurrent Builds

jtnord commented:
:+1:

oleg-nenashev commented:
:+1: for the fix, the refactoring (if applicable) may be done later
